iree/samples/static_library/static_library_demo.c

```diff
@@ -0,0 +1,210 @@
+// Copyright 2021 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+// A example of static library loading in IREE. See the README.md for more info.
+// Note: this demo requires artifacts from iree-translate before it will run.
+
+#include <stdio.h>
+
+#include "iree/base/api.h"
+#include "iree/hal/api.h"
+#include "iree/hal/local/executable_loader.h"
+#include "iree/hal/local/loaders/static_library_loader.h"
+#include "iree/hal/local/task_device.h"
+#include "iree/modules/hal/hal_module.h"
+#include "iree/samples/static_library/static_library_module.h"
+#include "iree/task/api.h"
+#include "iree/vm/api.h"
+#include "iree/vm/bytecode_module.h"
+
+// Compiled static library module here to avoid IO:
+#include "iree/samples/static_library/static_module.h"
+
+// A function to create the HAL device from the different backend targets.
+// The HAL device is returned based on the implementation, and it must be
+// released by the caller.
+iree_status_t create_sample_device(iree_hal_device_t** device) {
+  // Set paramters for the device created in the next step.
+  iree_hal_task_device_params_t params;
+  iree_hal_task_device_params_initialize(&params);
+
+  // Load the statically embedded library
+  const iree_hal_executable_library_header_t** static_library =
+      simple_mul_dispatch_0_library_query(
+          IREE_HAL_EXECUTABLE_LIBRARY_LATEST_VERSION, /*reserved=*/NULL);
+  const int library_count = 1;
+  const iree_hal_executable_library_header_t** libraries[1] = {static_library};
+
+  iree_hal_executable_loader_t* loaders[1] = {NULL};
+  iree_host_size_t loader_count = 0;
+  IREE_RETURN_IF_ERROR(iree_hal_static_library_loader_create(
+      1, libraries, iree_allocator_system(), &loaders[loader_count++]));
+
+  iree_task_executor_t* executor = NULL;
+  IREE_RETURN_IF_ERROR(
+      iree_task_executor_create_from_flags(iree_allocator_system(), &executor));
+
+  iree_string_view_t identifier = iree_make_cstring_view("dylib");
+
+  // Create the device and release the executor and loader afterwards.
+  IREE_RETURN_IF_ERROR(iree_hal_task_device_create(
+      identifier, &params, executor, IREE_ARRAYSIZE(loaders), loaders,
+      iree_allocator_system(), device));
+  iree_task_executor_release(executor);
+  for (iree_host_size_t i = 0; i < loader_count; ++i) {
```

Comment:
with the simplification above this can just be `iree_hal_executable_loader_release(library_loader);`